### PR TITLE
Moved assembleReads() to the util class

### PR DESCRIPTION
**Pull request by takutosato**
To be shared between HaplotypeCaller and Mutect2.

### CODE REVIEW

**Comment by takutosato** [comment]
@davidbenjamin could you review this small PR?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java [hunk 1]
```diff
@@ -215,4 +218,55 @@ public static VariantContext makeMergedVariantContext(final List<VariantContext>
...
+                                                  final SampleList sampleList,
+                                                  final Logger logger,
+                                                  final CachingIndexedFastaSequenceFile referenceReader,
+                                                  final ReadThreadingAssembler assemblyEngine){
+        finalizeRegion(region, argumentCollection.errorCorrectReads, argumentCollection.dontUseSoftClippedBases, (byte)(argumentCollection.minBaseQualityScore - 1), header, sampleList);
+        if( argumentCollection.debug) {
+            logger.info("Assembling " + region.getSpan() + " with " + region.size() + " reads:    (with overlap region = " + region.getExtendedSpan() + ")");
+        }
+
+        final int REFERENCE_PADDING = 500;
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java]: Make this a class-level static constant.

Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java]: And call it `REFERENCE_PADDING_FOR_ASSEMBLY`.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java [hunk 2]
```diff
@@ -215,4 +218,55 @@ public static VariantContext makeMergedVariantContext(final List<VariantContext>
...
+        final byte[] fullReferenceWithPadding = region.getAssemblyRegionReference(referenceReader, REFERENCE_PADDING);
+        final SimpleInterval paddedReferenceLoc = getPaddedReferenceLoc(region, REFERENCE_PADDING, referenceReader);
+        final Haplotype referenceHaplotype = createReferenceHaplotype(region, paddedReferenceLoc, referenceReader);
+
+        final ReadErrorCorrector readErrorCorrector = argumentCollection.errorCorrectReads ?
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java]: Thank you for the ternary.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java [hunk 3]
```diff
@@ -215,4 +218,55 @@ public static VariantContext makeMergedVariantContext(final List<VariantContext>
...
+                new ReadErrorCorrector(
+                        argumentCollection.assemblerArgs.kmerLengthForReadErrorCorrection,
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java]: The first argument doesn't need to be on a new line.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java [hunk 4]
```diff
@@ -470,7 +467,7 @@ public ActivityProfileState isActive( final AlignmentContext context, final Refe
         }
 
         // run the local assembler, getting back a collection of information on how we should proceed
-        final AssemblyResultSet untrimmedAssemblyResult = assembleReads(region, givenAlleles);
+        final AssemblyResultSet untrimmedAssemblyResult =  AssemblyBasedCallerUtils.assembleReads(region, Collections.emptyList(), hcArgs, readsHeader, samplesList, logger, referenceReader, assemblyEngine);
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java]: Shouldn't the second argument be `givenAlleles`?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java [hunk 5]
```diff
@@ -53,6 +55,7 @@
         return result;
     }
 
+
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java]: extra whitespace line

**Comment by davidbenjamin** [comment]
@takutosato Minor comments.  I went through this carefully  (the integration tests for `HaplotypeCaller` don't cover GGA mode) and am convinced that you didn't change behavior).

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java (hunk 5 above)
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java (hunk 1 above)
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java (hunk 3 above)
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java (hunk 4 above)
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java]: thanks, good catch

**Comment by takutosato** [comment]
Thanks for reviewing @davidbenjamin. I made the changes you suggested.

**Comment by davidbenjamin** [comment]
@takutosato Good to go.  Squash and merge.

**Pull request merged**